internal/metamorphic/crossversion/crossversion_test: TestMetaCrossVersion failed

internal/metamorphic/crossversion/crossversion_test.TestMetaCrossVersion [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicCrossversion/12621822?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicCrossversion/12621822?buildTab=artifacts#/) on refs/heads/master @ [b224e8b90a87](https://github.com/cockroachdb/pebble/commits/b224e8b90a87):


```
        iter61.SetBounds("hdmugvzfsd", "hdnlqwgidg")
        iter61.SeekLT("hdnlqwgidg", "")
        iter61.Next("")
        iter61.Next("")
        iter61.Prev("")
        iter61.Next("")
        iter61.Prev("uveyl@43")
        snap43.Get("feqwpzjbep@127")
        iter61.SeekLT("gllg@147", "")
        iter61.First()
        iter63 = snap43.NewIter("nfnopmwhky", "nxwlrrsnqb", 1 /* key types */, 0, 0, false /* use L6 filters */, "" /* masking suffix */)
        db1.RangeKeySet("gicubw", "jpbqdsxecq", "@94", "aykfggsdfgwrynhl")
        iter61.Next("")
        snap43.Get("nrwunep@68")
        iter63.Prev("")
        iter61.NextPrefix()
        iter63.SetBounds("dszce", "etjigkoh")
        iter63.SeekGE("dszce", "")
        iter63.Next("")
        db1.Set("vrlqynizfhd@151", "kfjujxqnqe")
        iter61.Prev("")
        iter61.SeekPrefixGE("feqwpzjbep@21")
        iter63.SeekPrefixGE("gvovvtphcl@62")
        db1.Flush()
        iter61.Next("")
        iter63.First()
        iter61.Last()
        iter63.Last()
        iter61.Close()
        iter63.Close()
        snap43.Close()
        db1.Get("gvovvtphcl@137")
        db1.Set("feqwpzjbep@80", "phpmletwiv")
        db1.Get("hdnls@53")
        db1.Get("feqwpzjbep@60")
        db1.Get("gvovvtphcl@83")
        db1.Get("uveyl@22")
        db1.Get("gvovvtphcl@76")
        db1.Set("gllg@73", "aeqiaypaokxdxwgj")
        db1.Set("agjw@23", "edxybwcgy")
        db1.Get("uveyl@15")
        db1.Set("feqwpzjbep@82", "bdlkdgdl")
        db1.Get("uveyl@150")
        snap45 = db1.NewSnapshot()
        db1.Get("uveyl@75")
        snap45.Close()
        db1.Close()
        
        		
        --- FAIL: TestMetaCrossVersion/execution-master/sha=043c3795-seed=7144057419648618075-opts=random-022(sha=db91e5c9-seed=3808559344755705723-opts=random-001(sha=ffd5ce9c-seed=7809109350420410039-opts=standard-000(sha=4ddacdaf-seed=12929196215159896102-opts=random-014(<empty>)))) (366.41s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/test-eng
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaCrossVersion.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

internal/metamorphic/crossversion/crossversion_test.TestMetaCrossVersion [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicCrossversion/12634826?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicCrossversion/12634826?buildTab=artifacts#/) on refs/heads/master @ [b224e8b90a87](https://github.com/cockroachdb/pebble/commits/b224e8b90a87):


```
        iter65.SeekLT("vizbangpcic@12", "")
        db1.Get("tkqphpsoejz@16")
        db1.Delete("uahpjso@11")
        db1.Merge("ukhkub@14", "mzdaveymkyvk")
        db1.Set("fdkizek@4", "cwfvqgnbifs")
        iter65.Prev("")
        iter65.SetBounds("aaexaiid@2", "aaexaiid@2")
        iter65.SeekLT("aaexaiid@2", "")
        iter65.Next("")
        iter65.Prev("")
        iter67.Next("")
        iter65.Last()
        iter65.Prev("")
        db1.Delete("ezlopenkcq@7")
        iter65.SeekPrefixGE("atltm@7")
        db1.Merge("phydoq@5", "zdd")
        snap50 = db1.NewSnapshot("bocavcrmjsq", "ezlopenkcq", "jymx", "njvhubw", "tplyrk", "vhnaccdjipy")
        iter67.SeekGE("oksqoypxhpb@2", "")
        iter65.SetBounds("aaexaiid@2", "aaexaiid@2")
        iter65.SeekLT("aaexaiid@2", "")
        iter65.Next("")
        iter65.Prev("")
        db1.SingleDelete("wxjex@15", false /* maybeReplaceDelete */)
        iter65.SeekLT("edbpajs@8", "")
        db1.RangeKeyUnset("lyiqfqp", "ounrxu", "@8")
        db1.Merge("jcqvbnotqm@3", "mudwnevhspsxpkie")
        iter65.Next("")
        db1.Merge("hteeyr@11", "ybtjwj")
        iter65.SetBounds("aaexaiid@2", "aaexaiid@2")
        iter65.SeekGE("aaexaiid@2", "")
        iter65.Prev("")
        db1.Merge("pdfaijgfx@16", "qzvvigv")
        db1.Merge("mjspw@4", "ofjlrphagucpqo")
        db1.Set("wdugfo@11", "ngxucotztpyghptboe")
        iter65.Last()
        iter65.Next("")
        db1.Delete("brkveh@6")
        iter65.Prev("")
        iter65.SeekLT("arnwtrsw@9", "")
        iter67.SetBounds("jnncw@2", "nzmsdsnfv@2")
        iter67.SeekLT("nzmsdsnfv@2", "")
        db1.Merge("vninwzzyyg@10", "kjcmcigiexlf")
        db1.Set("amtsaj@2", "qgkjhmzb")
        iter67.Close()
        iter65.Close()
        snap50.Close()
        db1.Close()
        
        		
        --- FAIL: TestMetaCrossVersion/execution-master/sha=043c3795-seed=17069796130237528405-opts=standard-015(sha=db91e5c9-seed=4809650095474515633-opts=random-006(sha=ffd5ce9c-seed=9181115102268245349-opts=random-002(sha=4ddacdaf-seed=13024630639060741452-opts=standard-021(<empty>)))) (389.84s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaCrossVersion.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

internal/metamorphic/crossversion/crossversion_test.TestMetaCrossVersion [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicCrossversion/12647564?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicCrossversion/12647564?buildTab=artifacts#/) on refs/heads/master @ [b224e8b90a87](https://github.com/cockroachdb/pebble/commits/b224e8b90a87):


```
        iter63.SeekLT("ofqa@44", "doyxdndd@124")
        iter62.SetOptions("", "uhmlul@121", 2 /* key types */, 0, 0, false /* use L6 filters */, "" /* masking suffix */)
        iter62.Last()
        db1.Set("rdhwkurym@161", "lnykqztcdlcpowdnuxs")
        iter61.Last()
        iter62.Prev("rdhwkurym@102")
        db1.Set("ofqa@89", "zwtvkdjeqzwlpqtktc")
        iter61.Prev("clts@13")
        iter63.SeekPrefixGE("clts@152")
        iter63.Prev("")
        iter62.SetBounds("uhmlul@37", "umlmtv@54")
        iter62.SeekLT("umlmtv@54", "")
        iter62.Next("")
        iter61.Prev("rdhwkurym@73")
        iter61.SeekGE("rdhwkurym@102", "")
        iter64 = db1.NewIter("", "", 2 /* key types */, 0, 0, false /* use L6 filters */, "@52" /* masking suffix */)
        iter64.SeekPrefixGE("idhwfst@130")
        db1.Get("rdhwkurym@21")
        db1.Set("vkneelp@159", "ygnmkdbplycm")
        db1.RangeKeySet("idhwfst", "pxaybajgjo", "@161", "kvsnecs")
        iter61.SetBounds("clts@25", "dnhbtpnlxya@57")
        iter61.SeekLT("dnhbtpnlxya@57", "")
        iter61.Next("")
        iter61.Prev("")
        iter63.SeekLT("umlmtv@75", "")
        iter64.First()
        iter62.NextPrefix()
        db1.Get("clts@127")
        iter62.SeekLT("ofqa@18", "")
        iter61.Prev("")
        db1.Get("uhmlul@123")
        iter61.First()
        iter63.SetBounds("clts@12", "doyxdndd@107")
        iter63.SeekLT("doyxdndd@107", "")
        iter63.Prev("")
        iter61.First()
        iter63.SeekGE("eujltfe@103", "")
        iter61.SeekLT("clts@36", "")
        iter64.SeekPrefixGE("clts@53")
        iter63.Last()
        iter61.SeekGE("eujltfe@102", "umlmtv@64")
        iter62.SeekGE("rdhwkurym@30", "")
        iter61.Close()
        iter64.Close()
        iter62.Close()
        iter63.Close()
        db1.Close()
        
        		
        --- FAIL: TestMetaCrossVersion/execution-master/sha=043c3795-seed=5825117174656947765-opts=random-006(sha=db91e5c9-seed=15674680038571973183-opts=random-000(sha=ffd5ce9c-seed=14294268487205330014-opts=standard-010(sha=4ddacdaf-seed=10904228769958173713-opts=random-005(<empty>)))) (445.29s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaCrossVersion.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/batch.go
+++ b/batch.go
@@ -659,31 +659,38 @@ func (b *Batch) prepareDeferredKeyRecord(keyLen int, kind InternalKeyKind) {
 	b.data = b.data[:pos+keyLen]
 }
 
-// AddInternalKey allows the caller to add an internal key of point key kinds to
-// a batch. Passing in an internal key of kind RangeKey* or RangeDelete will
-// result in a panic. Note that the seqnum in the internal key is effectively
-// ignored, even though the Kind is preserved. This is because the batch format
-// does not allow for a per-key seqnum to be specified, only a batch-wide one.
+// AddInternalKey allows the caller to add an internal key of point key or range
+// key kinds (but not RangeDelete) to a batch. Passing in an internal key of
+// kind RangeDelete will result in a panic. Note that the seqnum in the internal
+// key is effectively ignored, even though the Kind is preserved. This is
+// because the batch format does not allow for a per-key seqnum to be specified,
+// only a batch-wide one.
 //
 // Note that non-indexed keys (IngestKeyKind{LogData,IngestSST}) are not
 // supported with this method as they require specialized logic.
 func (b *Batch) AddInternalKey(key *base.InternalKey, value []byte, _ *WriteOptions) error {
 	keyLen := len(key.UserKey)
 	hasValue := false
-	switch key.Kind() {
-	case InternalKeyKindRangeDelete, InternalKeyKindRangeKeySet, InternalKeyKindRangeKeyUnset, InternalKeyKindRangeKeyDelete:
-		panic("unexpected range delete or range key kind in AddInternalKey")
+	switch kind := key.Kind(); kind {
+	case InternalKeyKindRangeDelete:
+		panic("unexpected range delete in AddInternalKey")
 	case InternalKeyKindSingleDelete, InternalKeyKindDelete:
-		b.prepareDeferredKeyRecord(len(key.UserKey), key.Kind())
-	default:
-		b.prepareDeferredKeyValueRecord(keyLen, len(value), key.Kind())
+		b.prepareDeferredKeyRecord(keyLen, kind)
+		b.deferredOp.index = b.index
+	case InternalKeyKindRangeKeySet, InternalKeyKindRangeKeyUnset, InternalKeyKindRangeKeyDelete:
+		b.prepareDeferredKeyValueRecord(keyLen, len(value), kind)
 		hasValue = true
+		b.incrementRangeKeysCount()
+	default:
+		b.prepareDeferredKeyValueRecord(keyLen, len(value), kind)
+		hasValue = true
+		b.deferredOp.index = b.index
 	}
-	b.deferredOp.index = b.index
 	copy(b.deferredOp.Key, key.UserKey)
 	if hasValue {
 		copy(b.deferredOp.Value, value)
 	}
+
 	// TODO(peter): Manually inline DeferredBatchOp.Finish(). Mid-stack inlining
 	// in go1.13 will remove the need for this.
 	if b.index != nil {

--- a/batch_test.go
+++ b/batch_test.go
@@ -45,8 +45,7 @@ func testBatch(t *testing.T, size int) {
 
 		for _, tc := range testCases {
 			if indexedPointKindsOnly && (tc.kind == InternalKeyKindLogData || tc.kind == InternalKeyKindIngestSST ||
-				tc.kind == InternalKeyKindRangeKeyUnset || tc.kind == InternalKeyKindRangeKeySet ||
-				tc.kind == InternalKeyKindRangeKeyDelete || tc.kind == InternalKeyKindRangeDelete) {
+				tc.kind == InternalKeyKindRangeDelete) {
 				continue
 			}
 			kind, k, v, ok := r.Next()
@@ -186,8 +185,7 @@ func testBatch(t *testing.T, size int) {
 	// Kind-specific methods.
 	for _, tc := range testCases {
 		if tc.kind == InternalKeyKindLogData || tc.kind == InternalKeyKindIngestSST ||
-			tc.kind == InternalKeyKindRangeKeyUnset || tc.kind == InternalKeyKindRangeKeySet ||
-			tc.kind == InternalKeyKindRangeKeyDelete || tc.kind == InternalKeyKindRangeDelete {
+			tc.kind == InternalKeyKindRangeDelete {
 			continue
 		}
 		key := []byte(tc.key)

--- a/metamorphic/ops.go
+++ b/metamorphic/ops.go
@@ -484,7 +484,7 @@ func (o *ingestOp) run(t *test, h historyRecorder) {
 		b := t.getBatch(id)
 		iter, rangeDelIter, rangeKeyIter := private.BatchSort(b)
 		db := t.getDB(o.dbID)
-		c, err := o.collapseBatch(t, db, iter, rangeDelIter, rangeKeyIter)
+		c, err := o.collapseBatch(t, db, iter, rangeDelIter, rangeKeyIter, b)
 		if err == nil {
 			err = db.Apply(c, t.writeOpts)
 		}
@@ -580,18 +580,29 @@ func (o *ingestOp) build(t *test, h historyRecorder, b *pebble.Batch, i int) (st
 
 	if rangeKeyIter != nil {
 		for span := rangeKeyIter.First(); span != nil; span = rangeKeyIter.Next() {
-			// Zero the sequence numbers of keys in this span.
-			zeroedSpan := &keyspan.Span{
+			// Coalesce the keys of this span and then zero the sequence
+			// numbers. This is necessary in order to make the range keys within
+			// the ingested sstable internally consistent at the sequence number
+			// it's ingested at. The individual keys within a batch are
+			// committed at unique sequence numbers, whereas all the keys of an
+			// ingested sstable are given the same sequence number. A span
+			// contaning keys that both set and unset the same suffix at the
+			// same sequence number is nonsensical, so we "coalesce" or collapse
+			// the keys.
+			collapsed := keyspan.Span{
 				Start: span.Start,
 				End:   span.End,
-				Keys:  make([]keyspan.Key, len(span.Keys)),
+				Keys:  make([]keyspan.Key, 0, len(span.Keys)),
 			}
-			for i := range span.Keys {
-				zeroedSpan.Keys[i] = span.Keys[i]
-				zeroedSpan.Keys[i].Trailer = base.MakeTrailer(0, span.Keys[i].Kind())
+			err = rangekey.Coalesce(t.opts.Comparer.Compare, equal, span.Keys, &collapsed.Keys)
+			if err != nil {
+				return "", err
 			}
-			keyspan.SortKeysByTrailer(&zeroedSpan.Keys)
-			if err := rangekey.Encode(zeroedSpan, w.AddRangeKey); err != nil {
+			for i := range collapsed.Keys {
+				collapsed.Keys[i].Trailer = base.MakeTrailer(0, collapsed.Keys[i].Kind())
+			}
+			keyspan.SortKeysByTrailer(&collapsed.Keys)
+			if err := rangekey.Encode(&collapsed, w.AddRangeKey); err != nil {
 				return "", err
 			}
 		}
@@ -643,6 +654,7 @@ func (o *ingestOp) collapseBatch(
 	db *pebble.DB,
 	pointIter base.InternalIterator,
 	rangeDelIter, rangeKeyIter keyspan.FragmentIterator,
+	b *pebble.Batch,
 ) (*pebble.Batch, error) {
 	defer closeIters(pointIter, rangeDelIter, rangeKeyIter)
 	equal := t.opts.Comparer.Equal
@@ -667,6 +679,12 @@ func (o *ingestOp) collapseBatch(
 		var lastUserKey []byte
 		for key, value := pointIter.First(); key != nil; key, value = pointIter.Next() {
 			// Ignore duplicate keys.
+			//
+			// Note: this is necessary due to MERGE keys, otherwise it would be
+			// fine to include all the keys in the batch and let the normal
+			// sequence number precedence determine which of the keys "wins".
+			// But the code to build the ingested sstable will only keep the
+			// most recent internal key and will not merge across internal keys.
 			if equal(lastUserKey, key.UserKey) {
 				continue
 			}
@@ -709,27 +727,28 @@ func (o *ingestOp) collapseBatch(
 		pointIter = nil
 	}
 
+	// There's no equivalent of a MERGE operator for range keys, so there's no
+	// need to collapse the range keys here. Rather than reading the range keys
+	// from `rangeKeyIter`, which will already be fragmented, read the range
+	// keys from the batch and copy them verbatim. This marginally improves our
+	// test coverage over the alternative approach of pre-fragmenting and
+	// pre-coalescing before writing to the batch.
+	//
+	// The `rangeKeyIter` is used only to determine if there are any range keys
+	// in the batch at all, and only because we already have it handy from
+	// private.BatchSort.
 	if rangeKeyIter != nil {
-		for span := rangeKeyIter.First(); span != nil; span = rangeKeyIter.Next() {
-			for i := range span.Keys {
-				var err error
-				switch span.Keys[i].Kind() {
-				case pebble.InternalKeyKindRangeKeyDelete:
-					err = collapsed.RangeKeyDelete(span.Start, span.End, t.writeOpts)
-				case pebble.InternalKeyKindRangeKeySet:
-					err = collapsed.RangeKeySet(span.Start, span.End, span.Keys[i].Suffix, span.Keys[i].Value, t.writeOpts)
-				case pebble.InternalKeyKindRangeKeyUnset:
-					err = collapsed.RangeKeyUnset(span.Start, span.End, span.Keys[i].Suffix, t.writeOpts)
-				default:
-					err = errors.Errorf("unknown batch range key record kind: %d", span.Keys[i].Kind())
-				}
-				if err != nil {
-					return nil, err
-				}
+		for r := b.Reader(); ; {
+			kind, key, value, ok := r.Next()
+			if !ok {
+				break
+			} else if !rangekey.IsRangeKey(kind) {
+				continue
 			}
-		}
-		if err := rangeKeyIter.Error(); err != nil {
-			return nil, err
+			ik := base.MakeInternalKey(key, 0, kind)
+			if err := collapsed.AddInternalKey(&ik, value, nil); err != nil {
+				return nil, err
+			}
 		}
 		if err := rangeKeyIter.Close(); err != nil {
 			return nil, err

